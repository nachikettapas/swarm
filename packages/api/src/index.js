// @flow

import regeneratorRuntime from 'regenerator-runtime';
import hash from 'object-hash';

import Client from '@swarm/client';
import Op, { Frame, ZERO, UUID, FRAME_SEP, js2ron } from '@swarm/ron';
import { ZERO as ZERO_UUID } from '@swarm/ron-uuid';
import { lww, set, ron2js } from '@swarm/rdt';
import type { Atom } from '@swarm/ron';
import type { Options as ClntOpts } from '@swarm/client';

export type Options = ClntOpts & {
  gcPeriod?: number,
  strictMode?: boolean,
};

export type Value = { [string]: Atom | Value } | Value[] | null;

interface Subscription {
  off(): boolean;
  is(hash: string): boolean;
}

export default class API {
  client: Client;
  options: Options;
  subs: Array<Subscription>;
  cache: { [string]: { [string]: Atom } };
  gcInterval: IntervalID;

  constructor(options: Options): API {
    this.client = new Client(options);
    this.options = options;
    this.subs = [];
    this.cache = {};
    if (options.gcPeriod) {
      this.gcInterval = setInterval(this.gc.bind(this), options.gcPeriod);
    }
    // $FlowFixMe
    this.uuid = this.uuid.bind(this);
    return this;
  }

  ensure(): Promise<void> {
    return this.client.ensure();
  }

  uuid(): UUID {
    if (!this.client.clock)
      throw new Error(
        'have no clock yet, invoke `await <swarm>.ensure()` first',
      );
    return this.client.clock.time();
  }

  // garbage collection for unused cached data
  gc(): {| deleted: number, existing: number |} {
    const ret = { deleted: 0, existing: 0 };
    const lstnrs = {};
    for (const id of Object.keys(this.client.lstn)) lstnrs[id] = true;

    for (const id of Object.keys(this.cache)) {
      if (!lstnrs[id]) {
        delete this.cache[id];
        ret.deleted++;
      } else {
        ret.existing++;
      }
    }

    return ret;
  }

  async set(
    id: string | UUID,
    value: { [string]: Atom | void },
  ): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    await this.client.ensure();

    if (this.options.strictMode) {
      const type = await this.typeOf(id);
      if (type && type !== lww.type.toString()) return false;
    }

    const frame = new Frame();

    frame.push(
      new Op(lww.type, uuid, this.uuid(), ZERO_UUID, undefined, FRAME_SEP),
    );

    for (const k of Object.keys(value).sort()) {
      const op = new Op(
        lww.type,
        uuid,
        frame.last.uuid(2),
        UUID.fromString(k),
        undefined,
        ',',
      );

      if (value[k] !== undefined) {
        op.values = js2ron([value[k]]);
        if (!uuid.isLocal() && value[k] instanceof UUID && value[k].isLocal()) {
          return false;
        }
      }

      frame.push(op);
    }

    if (frame.isPayload()) {
      await this.client.push(frame.toString());
      return true;
    }
    return false;
  }

  async add(id: string | UUID, value: Atom): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    await this.client.ensure();

    if (this.options.strictMode) {
      const type = await this.typeOf(id);
      if (type && type !== set.type.toString()) return false;
    }

    const frame = new Frame();
    const time = this.uuid();
    let op = new Op(set.type, uuid, time, ZERO_UUID, undefined, FRAME_SEP);
    frame.push(op);
    op = op.clone();

    if (!uuid.isLocal() && value instanceof UUID && value.isLocal()) {
      return false;
    }

    op.values = js2ron([value]);
    frame.pushWithTerm(op, ',');

    await this.client.push(frame.toString());
    return true;
  }

  async remove(id: string | UUID, value: Atom): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    id = uuid.toString();
    await this.client.ensure();

    if (this.options.strictMode) {
      const type = await this.typeOf(id);
      if (type !== set.type.toString()) return false;
    }

    const frame = new Frame();
    const ts = this.uuid();
    let deleted = false;
    let op = new Op(set.type, uuid, ts, ZERO_UUID, undefined, FRAME_SEP);
    frame.push(op);

    let state = await this.client.storage.get(id);
    if (!state) return false;

    const str = js2ron([value]);
    for (const v of new Frame(state)) {
      if (!v.isRegular()) continue;
      if (v.values === str) {
        deleted = true;
        op = op.clone();
        op.location = v.event;
        op.values = '';
        frame.pushWithTerm(op, ',');
      }
    }

    if (deleted) {
      await this.client.push(frame.toString());
    }
    return deleted;
  }

  close(): Promise<void> {
    return this.client.close();
  }

  open(): void {
    return this.client.open();
  }

  async typeOf(id: string | UUID): Promise<string | null> {
    const obj = this.cache[id.toString()];
    if (obj !== undefined) {
      // found in cache
      return obj && obj.type && typeof obj.type === 'string' ? obj.type : '';
    }

    const state = await this.client.storage.get(id.toString());
    if (state) {
      const op = Op.fromString(state);
      if (op) return op.uuid(0).toString();
    }
    // type is not defined
    return null;
  }
}
