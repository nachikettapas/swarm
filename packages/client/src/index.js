/* @flow */

import regeneratorRuntime from 'regenerator-runtime'; // for async/await work flow

import RWS from './rws';

import Op, {
  Batch,
  Frame,
  Cursor,
  UUID,
  QUERY_SEP,
  FRAME_SEP,
  mapUUIDs,
  js2ron,
} from '@swarm/ron';
import type { Atom } from '@swarm/ron';
import type { Clock } from '@swarm/clock';
import { Logical, Calendar } from '@swarm/clock';
import { ZERO, NEVER, LOCAL } from '@swarm/ron-uuid';
import { lww } from '@swarm/rdt';
import { reduce } from '@swarm/rdt';
import type { Storage } from './storage';
import { InMemory } from './storage';
import type { Connection } from './connection';
import Pending from './pending';

export { default as WebSocket } from './rws';
export { InMemory, LocalStorage } from './storage';
export type { Connection } from './connection';
import { DevNull } from './connection';

export type Meta = {
  id?: string,
  name: string,
  clockMode?: 'Logical' | 'Calendar',
  clockLen?: number,
  forkMode?: string,
  peerIdBits?: number,
  horizont?: number,
  auth?: string,
  seen?: string,
  offset?: number,
  '0'?: number,
};

export type Options = {
  storage: Storage,
  upstream?: string | Connection,
  hsTimeout?: number,
  fetchTimeout?: number,
  resendAfter?: number,
  db?: Meta,
};

type callback = {
  f: (frame: string, state: string | null) => void,
  once?: true,
  ensure?: true,
};

const defaultMeta: Meta = {
  name: 'default',
  // FIXME clockMode: 'Calendar',
  clockLen: 5,
  forkMode: '// FIXME', // keep possible values sync with server
  peerIdBits: 30,
  horizont: 604800, // one week in seconds
  offset: 0,
};

const SEEN = '0';
const META = '__meta__';

// A bare-bone swarm client. Consumes updates from the server,
// feeds resulting RON states back to the listeners.
export default class Client {
  clock: ?Clock;
  lstn: { [key: string]: Array<callback> };
  upstream: Connection;
  storage: Storage;
  queue: Array<[() => void, (err: Error) => void]> | void;
  db: Meta;
  options: {
    hsTimeout: number,
    fetchTimeout: number,
    resendAfter: number,
  };
  pending: Pending;
  seen: UUID;

  constructor(options: Options): Client {
    this.db = {
      ...defaultMeta,
      ...options.db,
    };
    this.storage = options.storage;
    this.lstn = {};
    this.queue = [];
    this.options = {
      hsTimeout: options.hsTimeout || 3e5 /* 5min */,
      fetchTimeout: options.fetchTimeout || 3e4 /* 30sec */,
      resendAfter: options.resendAfter || 0,
    };
    this.init(options);
    this.seen = ZERO;
    return this;
  }

  async init(options: Options) {
    try {
      const { [META]: meta, [SEEN]: seen } = await this.storage.multiGet([
        META,
        SEEN,
      ]);
      const last = seen ? UUID.fromString(seen) : ZERO;

      this.db = ({
        ...this.db,
        ...JSON.parse(meta || '{}'),
      }: Meta);

      if (this.db.clockMode && this.db.id) {
        switch (this.db.clockMode) {
          case 'Logical':
            this.clock = new Logical(this.db.id, {
              length: this.db.clockLen,
              last,
            });
            break;
          case 'Calendar':
            this.clock = new Calendar(this.db.id, {
              offset: this.db.offset || 0,
              length: this.db.clockLen,
              last,
            });
            break;
          default:
            throw new Error(
              `TODO: Clock mode '${this.db.clockMode}' is not supported yet`,
            );
        }
        this.seen = last;
      }

      if (typeof options.upstream === 'string') {
        this.upstream = new RWS(options.upstream, undefined, {
          reconnectOnError: true,
        });
      } else if (options.upstream) {
        this.upstream = options.upstream;
      } else if (!options.upstream && this.db.clockMode) {
        this.upstream = new DevNull();
      } else if (!options.upstream && !this.db.clockMode) {
        throw new Error('neither connection options nor clock options found');
      }

      this.pending = await Pending.read(this.storage);

      // check if we start with an already existing replica
      if (meta && this.clock) {
        this.upstream.onopen = () =>
          this.handshake().catch(async e => {
            await this.close();
            this.panic(e);
          });
        this.release(null);
      } else {
        this.upstream.onopen = () =>
          this.handshake()
            .then(() => this.release(null))
            .catch(async e => {
              await this.close();
              this.release((e: Error));
            });
      }
    } catch (e) {
      this.release((e: Error));
    }
  }

  async handshake(): Promise<void> {
    if (this.upstream instanceof DevNull) {
      await this.storage.set(META, JSON.stringify(this.db));
      this.upstream.onmessage = (me: MessageEvent) => {
        this.onMessage(((me.data: any): string)).catch(this.panic);
      };
      return;
    }

    const hs = new Promise((resolve, reject) => {
      setTimeout(reject, this.options.hsTimeout);
      this.upstream.onmessage = (me: MessageEvent) =>
        resolve(((me.data: any): string));
    });

    const hello = new Frame();
    const head = new Op(
      new UUID('db', '0', '$'),
      new UUID(this.db.name, '0', '$'),
      new UUID(this.seen.value, this.db.id || '0', '+'),
      ZERO,
      undefined,
      QUERY_SEP,
    );
    hello.push(head);
    hello.pushWithTerm(head, FRAME_SEP);

    const { auth } = this.db;
    if (auth) {
      hello.push(
        new Op(
          head.uuid(0),
          head.uuid(1),
          head.uuid(2),
          head.uuid(3),
          js2ron([auth]),
          ',',
        ),
      );
    }
    this.upstream.send(hello.toString());

    const resp = await hs;

    const dbOpts: Meta = {
      clockMode: 'Calendar',
      ...this.db,
    };

    let seen: UUID = ZERO;
    for (const op of new Frame(resp)) {
      if (op.uuid(3).isError()) {
        this.close();
        throw new Error(`Handshake error: ${op.uuid(3).toString()}`);
      }
      if (seen.eq(ZERO)) {
        seen = op.uuid(2);
      }
      const val = op.value(0);
      if (val) {
        let key = op.uuid(3).toString();
        key = key[0].toLowerCase() + key.slice(1);
        dbOpts[key] = val instanceof UUID ? val.value : val;
      }
    }

    const op = Op.fromString(resp);
    if (op) {
      dbOpts.id = op.uuid(1).origin;
    } else {
      throw new Error(
        `Expected replica id not found in a handshake response: \n\t'${(
          op || ''
        ).toString()}'`,
      );
    }

    if (this.clock) {
      if (this.db.clockMode !== dbOpts.clockMode)
        throw new Error(
          `Different clock mode: '${this.db.clockMode ||
            'undefined'}' !== '${dbOpts.clockMode || 'undefined'}'`,
        );
    } else if (dbOpts.id) {
      switch (dbOpts.clockMode) {
        case 'Logical':
          this.clock = new Logical(dbOpts.id, { length: dbOpts.clockLen });
          break;
        case 'Calendar':
          this.clock = new Calendar(dbOpts.id, { length: dbOpts.clockLen });
          break;
        default:
          throw new Error(
            `Clock mode '${dbOpts.clockMode ||
              'undefined'}' is not supported yet`,
          );
      }
    } else {
      throw new Error(
        `Clock mode '${dbOpts.clockMode || 'undefined'}' is not supported yet`,
      );
    }

    dbOpts.offset = this.clock.adjust(dbOpts['0'] || seen);
    delete dbOpts['0'];

    this.db = { ...this.db, ...dbOpts };
    await this.storage.set(META, JSON.stringify(this.db));
    this.upstream.onmessage = (me: MessageEvent) =>
      this.onMessage(((me.data: any): string)).catch(this.panic);
    this.upstream.onopen = () => this.handshake().catch(this.panic);

    const { resendAfter } = this.options;
    if (resendAfter) {
      this.pending.onIdle(this.onIdle);
      this.pending.setIdlePeriod(resendAfter);
    } else {
      // Resend all the frames
      for (const p of this.pending) this.upstream.send(p);
    }

    // Re-subscribe
    let query: Array<string> = Object.keys(this.lstn);
    if (query.length) {
      await this.on('#' + query.join('#'));
    }
  }

  // Ensure returns Promise which will be resolved after connection
  // installed or rejected if an error occurred.
  async ensure(): Promise<void> {
    const { queue } = this;
    if (queue === undefined) {
      return Promise.resolve();
    } else {
      return new Promise((release, reject) => {
        queue.push([release, reject]);
      });
    }
  }

  release(err: Error | null) {
    if (!this.queue) return;
    for (const cbk of this.queue) {
      if (err) {
        cbk[1](err);
      } else {
        cbk[0]();
      }
    }
    delete this.queue;
  }

  async close(): Promise<void> {
    if (this.upstream) this.upstream.close();
    this.pending.onIdle(() => {});
    this.pending.setIdlePeriod(0);
    await this.pending.flush();
  }

  open(): void {
    if (this.upstream) {
      this.upstream.open();
    }
  }

  async onMessage(message: string): Promise<void> {
    const { clock } = this;
    if (!clock) throw new Error('Have no clock');

    for (const op of new Frame(message)) {
      // check if it's an error from the server
      if (op.uuid(0).toString() === 'db') {
        this.close();
        // TODO destroy replica ?
        throw new Error(op.toString());
      }

      // skip malformed NEVERs explicitly
      // waiting for the fix at server side
      if (!op.event.eq(NEVER)) {
        await this.see(op.event);
        // handle ack
        if (op.event.origin === clock.origin()) {
          const seen = await this.pending.release(op.event);
        }
      }

      break;
    }

    for (const f of Batch.splitByID(message))
      await this.merge(f.toString(), { local: false });
  }

  // On installs subscriptions.
  async on(
    query: string,
    callback: ?(frame: string, state: string | null) => void,
    options: { once?: true, ensure?: true } = {},
  ): Promise<boolean> {
    await this.ensure();
    const fwd = new Frame();
    const upstrm = new Frame();
    const wrapped = { f: callback, ...options };
    let onceSent = 0;

    const keys = new Frame(query).unzip().map(op => {
      if (op.uuid(1).eq(ZERO))
        throw new Error(`ID is not specified: "${op.toString()}"`);
      return op.uuid(1).toString();
    });

    const data = await this.storage.multiGet(keys.concat('0'));

    for (const key of keys) {
      const id = UUID.fromString(key);
      let base = ZERO;
      const stored = data[key];

      // try to avoid network request
      if (callback && wrapped.once && (stored !== null || !wrapped.ensure)) {
        callback('#' + key, stored);
        onceSent++;
      } else {
        if (stored) {
          for (const op of new Frame(stored)) {
            base = op.event.lt(this.seen) ? op.event : this.seen;
            // base = op.event;
            break;
          }
        }

        let found = false;
        let exists = !!callback && !!this.lstn[key] && this.lstn[key].length;

        if (wrapped.f) {
          for (const l of this.lstn[key] || []) {
            found = found || l.f === wrapped.f;
          }

          if (!found) {
            // $FlowFixMe
            this.lstn[key] = (this.lstn[key] || []).concat([wrapped]);
          }
        }

        if (!found) {
          if (!exists && !id.isLocal()) {
            upstrm.push(
              new Op(ZERO, id, base, ZERO, '', QUERY_SEP + FRAME_SEP),
            ); // FIXME check fork mode
          }
          fwd.push(new Op(ZERO, id, base, ZERO, '', QUERY_SEP + FRAME_SEP)); // FIXME check fork mode
        }
      }
    }

    if (this.upstream && this.clock && upstrm.toString()) {
      this.upstream.send(upstrm.toString());
    }

    if (callback && fwd.toString()) {
      await this.notify(fwd.toString(), wrapped);
    }

    return !!fwd.toString() || !!onceSent;
  }

  // Off removes subscriptions.
  off(
    q: string | void,
    callback: ?(frame: string, state: string | null) => void,
  ): string | void {
    // unless query passed fetch all the keys to unsubscribe from them
    const query: string =
      q ||
      Object.keys(this.lstn)
        .map(i => '#' + i)
        .join('');
    let c = 0;
    const fwd = new Frame();
    fwd.push(new Op(ZERO, ZERO, NEVER, ZERO, undefined, QUERY_SEP));

    for (const op of new Frame(query)) {
      const key = op.uuid(1).toString();
      if (!this.lstn[key]) continue;
      if (callback) {
        this.lstn[key] = this.lstn[key].filter(cbk => cbk.f !== callback);
        if (!this.lstn[key].length) {
          if (!op.uuid(1).isLocal()) {
            fwd.push(new Op(op.type, op.object, NEVER, ZERO, undefined, ','));
            c++;
          }
          delete this.lstn[key];
        }
      } else {
        if (!op.uuid(1).isLocal()) {
          fwd.push(new Op(op.type, op.object, NEVER, ZERO, undefined, ','));
          c++;
        }
        delete this.lstn[key];
      }
    }
    if (!!c) {
      this.upstream.send(fwd.toString());
      return fwd.toString();
    }
    return;
  }

  // Push sends updates to remote and local storages.
  // Waits for connection installed. Thus, the client works in
  // read-only mode until installed connection.
  async push(rawFrame: string): Promise<void> {
    await this.ensure();
    let stamps: { [string]: UUID | void } = {};

    const frame = mapUUIDs(rawFrame, (uuid, position): UUID => {
      if (position === 0) return uuid.eq(ZERO) ? lww.type : uuid;
      if (position > 2 || !uuid.eq(ZERO)) return uuid;
      const exists = stamps[uuid.toString()];
      // $FlowFixMe
      return exists ? exists : (stamps[uuid.toString()] = this.clock.time());
    });

    const filtered = new Frame(frame)
      .filter(op => !op.uuid(1).isLocal())
      .toString();
    if (filtered) {
      await this.pending.push(filtered);
      this.upstream.send(filtered);
    }
    await this.merge(frame, { local: true });
  }

  // Notify calls back with an existing states
  async notify(
    frame: string,
    callback: {
      f: ?(frame: string, state: string | null) => void,
      once?: true,
      ensure?: true,
    },
  ): Promise<void> {
    if (!callback.f) return;
    const keys: { [string]: true } = {};
    for (const op of new Frame(frame)) keys[op.uuid(1).toString()] = true;
    const ks = Object.keys(keys);
    const store = await this.storage.multiGet(ks);
    for (const key of ks) {
      const value = store[key];
      const isLocal = key.slice(key.length - LOCAL.length) === LOCAL;
      if (!callback.ensure || value !== null || isLocal) {
        if (callback.once) this.off('#' + key, callback.f);
        // $FlowFixMe
        callback.f('#' + key, isLocal ? value || '' : value);
      }
    }
  }

  // Merge updates local state and notifies listeners.
  // Accepts single frame.
  //    notifies listeners with empty payload
  //    if there is no saved state
  async merge(
    frame: string,
    options: { local: boolean } = { local: true },
  ): Promise<void> {
    const fr = new Frame(frame);

    // Read only first operation of the frame
    // but with guarantee.
    for (const op of fr) {
      const key = op.uuid(1).toString();
      if (key === '0') return; // ack

      let notify: boolean = false;

      const updated = await this.storage.merge(key, (prev: string | null) => {
        if (fr.isPayload()) {
          if (typeof prev === 'string') {
            const update = reduce(
              Batch.fromStringArray(prev, frame),
            ).toString();
            notify = prev !== update;
            return update;
          }
          notify = true;
          return frame;
        }

        // An empty state from the server whereas a local state is full
        if (prev !== null) {
          return prev;
        }

        // Empty state from a server, hence, an object does not exist in the system.
        notify = true;
        return '';
      });

      if (notify) {
        // Copy an array to be able to unsubscribe before calling back
        for (const l of (this.lstn[key] || []).slice()) {
          if (!l.ensure || updated !== null) {
            if (l.once) this.off('#' + key, l.f);
            l.f('#' + key, updated);
          }
        }
      }
      break;
    }
  }

  panic = (err: any): void => {
    console.error('panic: %s', err);
  };

  onIdle = () => {
    // Resend all the frames
    for (const p of this.pending) this.upstream.send(p);
  };

  async see(event: UUID): Promise<UUID> {
    const { clock } = this;
    if (clock && clock.see(event)) {
      this.seen = event;
      await this.storage.set(SEEN, event.toString());
    }
    return event;
  }
}
