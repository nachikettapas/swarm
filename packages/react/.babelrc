{
  "presets": [
    ["env", {
      "loose": true,
      "useBuiltIns": false
    }],
    "react"
  ],
  "plugins": [
    ["transform-object-rest-spread", {
      "useBuiltIns": true }
    ],
    "transform-es2015-spread",
    "transform-flow-strip-types"
  ]
}
