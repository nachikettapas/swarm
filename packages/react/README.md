# Swarm + React

> bindings between Swarm and React

### Usage

Take a look at [test case](https://github.com/gritzko/swarm/blob/master/packages/react/__tests__/react.js)
