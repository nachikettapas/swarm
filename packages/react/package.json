{
  "name": "@swarm/react",
  "version": "0.1.2",
  "description": "Swarm to React bindings",
  "author": "Oleg Lebedev <ole6edev@gmail.com> (https://github.com/olebedev)",
  "main": "lib/index.js",
  "repository": {
    "type": "git",
    "url": ""
  },
  "dependencies": {
    "@swarm/api": "^0.1.1",
    "@swarm/db": "^0.1.2",
    "@swarm/ron": "^0.1.1",
    "@swarm/ron-uuid": "^0.1.1"
  },
  "devDependencies": {
    "babel-preset-react": "^6.24.1",
    "react": "16.2.0",
    "react-dom": "16.2.0",
    "react-test-renderer": "^16.2.0"
  },
  "peerDependencies": {
    "invariant": "^2.2.2",
    "prop-types": "^15.6.0",
    "react": "16.2.0",
    "react-dom": "16.2.0",
    "regenerator-runtime": "^0.11.1"
  },
  "files": [
    "lib/*.js",
    "lib/*.js.flow"
  ],
  "scripts": {
    "build": "yarn run build:clean && yarn run build:lib && yarn run build:flow",
    "build:clean": "../../node_modules/.bin/rimraf lib",
    "build:lib": "../../node_modules/.bin/babel -d lib src --ignore '**/__tests__/**'",
    "build:flow": "../../node_modules/.bin/flow-copy-source -v -i '**/__tests__/**' src lib"
  },
  "keywords": [
    "swarm",
    "react",
    "replicated",
    "RON",
    "CRDT"
  ],
  "publishConfig": {
    "access": "public"
  },
  "license": "MIT"
}
