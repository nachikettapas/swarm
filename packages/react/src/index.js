// @flow

export { default as Provider } from './Provider';
export { default as GraphQL } from './GraphQL';
export type { Response, Mutation } from './GraphQL';
