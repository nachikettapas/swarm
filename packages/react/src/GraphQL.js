// @flow

import regeneratorRuntime from 'regenerator-runtime'; // for async/await work flow

import * as React from 'react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import type { DocumentNode } from 'graphql';

import DB from '@swarm/db';
import type { Request, Response as DBResponse } from '@swarm/db';
import type { Value } from '@swarm/api';
import type { Variables } from '@swarm/db';
import type { Atom } from '@swarm/ron';
import UUID, { ERROR } from '@swarm/ron-uuid';

export type Mutation = Variables => Promise<Value>;

export type Response<T> = {
  data: ?T,
  uuid: ?() => UUID,
  error?: Error,
  mutations: { [string]: Mutation },
};

type Props<T> = {
  query?: DocumentNode,
  variables?: Variables,
  swarm?: DB,
  mutations?: { [string]: DocumentNode },
  children: (r: Response<T>) => React.Node,
};

type State<T> = {
  data: ?T,
  error?: Error,
  mutations: {
    [string]: Mutation,
  },
  initialized: boolean,
};

export default class GraphQL<T> extends React.Component<Props<T>, State<T>> {
  swarm: ?DB;
  _off: void | (() => boolean);
  _unmounted: boolean;

  constructor(props: Props<T>, context: { swarm: ?DB }) {
    super(props, context);
    this.swarm = context.swarm || this.props.swarm;
    invariant(
      this.swarm,
      `Could not find "swarm" in either the context or ` +
        `props of <GraphQL>. ` +
        `Either wrap the root component in a <Provider>, ` +
        `or explicitly pass "swarm" as a prop to <GraphQL>.`,
    );

    this._unmounted = false;
    this.state = {
      data: null,
      mutations: this._bindMutations(),
      initialized: this.swarm ? this.swarm.client.queue === undefined : false,
    };

    if (this.swarm) {
      this.swarm
        .ensure()
        .then(() => {
          this._subscribe();
          this.setState({ initialized: true });
        })
        .catch(error => {
          !this._unmounted && this.setState({ error });
        });
    }
  }

  componentDidUpdate(prev: Props<T>) {
    if (
      this.props.query !== prev.query ||
      !shallowEqual(this.props.variables, prev.variables) ||
      !shallowEqual(this.props.mutations, prev.mutations)
    ) {
      this._unsubscribe();
      !this._unmounted &&
        this.setState(
          {
            data: null,
            error: undefined,
            mutations: this._bindMutations(),
          },
          this._subscribe.bind(this),
        );
    }
  }

  componentWillUnmount() {
    this._unmounted = true;
    this._unsubscribe();
  }

  async _subscribe(): Promise<void> {
    const {
      props: { query, variables },
      swarm,
    } = this;
    if (!swarm || !swarm.execute || !query) return;

    const sub = await swarm.execute({ query, variables }, (r: DBResponse<any>) => {
      !this._unmounted && this.setState({ data: r.data, error: r.error });
    });

    if (this._off) this._off();
    this._off = sub.off;
  }

  _unsubscribe() {
    this._off && this._off();
  }

  _bindMutations(): {
    [string]: Mutation,
  } {
    const {
      props: { mutations },
      swarm,
    } = this;
    const ret = {};
    if (mutations && swarm) {
      for (const key of Object.keys(mutations)) {
        ret[key] = async (variables: Variables): Promise<Value> => {
          return new Promise((resolve, reject) => {
            swarm.execute({ query: mutations[key], variables }, (r: DBResponse<any>) => {
              r.error ? reject(r.error) : resolve(r.data);
            });
          });
        };
      }
    }
    return ret;
  }

  render() {
    return (
      this.props.children &&
      this.props.children.call(null, {
        data: this.state.data,
        uuid: this.state.initialized && this.swarm ? this.swarm.uuid : null,
        error: this.state.error,
        mutations: this.state.mutations,
      })
    );
  }
}

GraphQL.contextTypes = {
  swarm: PropTypes.shape({}).isRequired,
};

function shallowEqual(a?: {}, b?: {}): boolean {
  if (a === b) return true;
  if (a == null || b == null) return false;
  var aKeys = Object.keys(a);
  if (Object.keys(b).length !== aKeys.length) return false;
  for (const key of aKeys) {
    if (a[key] !== b[key]) return false;
  }
  return true;
}
