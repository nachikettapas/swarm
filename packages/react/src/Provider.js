// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import API from '@swarm/db';

export default class Provider extends React.Component<{
  swarm: API,
  children: React.ChildrenArray<React.Node>,
}> {
  getChildContext() {
    return { swarm: this.props.swarm };
  }

  render() {
    let { children } = this.props;
    return React.Children.only(children);
  }
}

Provider.childContextTypes = {
  swarm: PropTypes.shape({}).isRequired,
};
