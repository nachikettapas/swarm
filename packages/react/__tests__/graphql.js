// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';
import gql from 'graphql-tag';

import DB from '@swarm/db';
import { Provider, GraphQL } from '../src';
import type { Response } from '../src';
import { Connection } from '../../__tests__/fixtures';
import { InMemory } from '../../client/src/storage';

const Basic = ({ data, error, onClick }) => (
  <div>
    {JSON.stringify(data || error)}
    <button onClick={onClick}>update</button>
  </div>
);

const sub = gql`
  subscription test {
    object @node(id: "object") {
      id
      version
      test
      some
      additional
    }
  }
`;

const mutateObj = gql`
  mutation obj($payload: Payload!) {
    set(id: "object", payload: $payload)
  }
`;

test('React: graphql', async () => {
  const upstream = new Connection('018-react-graphql.ron');
  const storage = new InMemory();
  const api = new DB({
    storage,
    upstream,
    db: {
      id: 'user',
      name: 'test',
      auth: 'JwT.t0k.en',
      clockMode: 'Logical',
    },
  });

  const component = renderer.create(
    <Provider swarm={api}>
      <GraphQL query={sub} mutations={{ obj: mutateObj }}>
        {props => {
          return (
            <Basic
              {...props}
              onClick={(payload: Response<null>) => {
                // $FlowFixMe
                const p = payload || { test: props.data.object.test + 1 };
                if (props.mutations) {
                  // $FlowFixMe
                  props.mutations.obj({ payload: p });
                }
              }}
            />
          );
        }}
      </GraphQL>
    </Provider>,
  );

  await new Promise(r => setTimeout(r, 500));
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component.root.findByType('button').props.onClick({ some: 'value' });

  await new Promise(r => setTimeout(r, 100));

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component.root.findByType('button').props.onClick();

  await new Promise(r => setTimeout(r, 100));

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component.root.findByType('button').props.onClick({ additional: true });

  await new Promise(r => setTimeout(r, 100));

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const dump = upstream.dump();
  expect(dump.session).toEqual(dump.fixtures);
});
