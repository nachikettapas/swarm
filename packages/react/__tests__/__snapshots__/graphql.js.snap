// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React: graphql 1`] = `
<div>
  {"object":{"id":"object","version":"1ABC1+user","test":5,"some":null,"additional":null}}
  <button
    onClick={[Function]}
  >
    update
  </button>
</div>
`;

exports[`React: graphql 2`] = `
<div>
  {"object":{"id":"object","version":"1ABC1+user","test":5,"some":"value","additional":null}}
  <button
    onClick={[Function]}
  >
    update
  </button>
</div>
`;

exports[`React: graphql 3`] = `
<div>
  {"object":{"id":"object","version":"1ABC3+user","test":6,"some":"value","additional":null}}
  <button
    onClick={[Function]}
  >
    update
  </button>
</div>
`;

exports[`React: graphql 4`] = `
<div>
  {"object":{"id":"object","version":"1ABC3+user","test":6,"some":"value","additional":true}}
  <button
    onClick={[Function]}
  >
    update
  </button>
</div>
`;
